tss-esapi/src/abstraction/public.rs

```diff
@@ -2,127 +2,253 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use crate::interface_types::ecc::EccCurve;
-use crate::structures::{Public, RsaExponent};
+use crate::structures::Public;
+use crate::utils::PublicKey as TpmPublicKey;
 use crate::{Error, WrapperErrorKind};
 
 use core::convert::TryFrom;
-use oid::ObjectIdentifier;
-use picky_asn1::bit_string::BitString;
-use picky_asn1::wrapper::{IntegerAsn1, OctetStringAsn1};
-use picky_asn1_x509::{
-    AlgorithmIdentifier, EcParameters, EcPoint, PublicKey, RsaPublicKey, SubjectPublicKeyInfo,
+use elliptic_curve::{
+    generic_array::typenum::Unsigned,
+    sec1::{EncodedPoint, FromEncodedPoint, ModulusSize, ToEncodedPoint},
+    AffinePoint, CurveArithmetic, FieldBytesSize, PublicKey,
 };
 
-/// Can be converted from [`crate::structures::Public`] when not a fully constructed
-/// [`picky_asn1_x509::SubjectPublicKeyInfo`] is required.
-///
-/// # Details
-/// Holds either [`picky_asn1_x509::RsaPublicKey`] for [`crate::structures::Public::Rsa`] or
-/// [`picky_asn1_x509::EcPoint`] for [`crate::structures::Public::Ecc`].
-///
-/// This object can be serialized and deserialized
-/// using serde if the `serde` feature is enabled.
-#[derive(Debug, PartialEq, Eq, Clone)]
-#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
-pub enum DecodedKey {
-    RsaPublicKey(RsaPublicKey),
-    EcPoint(EcPoint),
+use x509_cert::spki::SubjectPublicKeyInfoOwned;
+
+#[cfg(all(feature = "rustcrypto", feature = "rsa"))]
+use {
+    crate::structures::RsaExponent,
+    rsa::{BigUint, RsaPublicKey},
+};
+
+#[cfg(all(
+    feature = "rustcrypto",
+    any(
+        feature = "p192",
+        feature = "p224",
+        feature = "p256",
+        feature = "p384",
+        feature = "p521",
+        feature = "rsa",
+        feature = "sm2"
+    )
+))]
+use pkcs8::EncodePublicKey;
+
+/// Default exponent for RSA keys.
+// Also known as 0x10001
+#[cfg(all(feature = "rustcrypto", feature = "rsa"))]
+const RSA_DEFAULT_EXP: u64 = 65537;
+
+impl<C> TryFrom<&Public> for PublicKey<C>
```

Comment:
Should this (and all the other similar `impl`s below) not be gated on the `rustcrypto` feature being enabled?

Reply:
no, because it's all gated by `abstraction` which pulls `rustcrypto`, but I can rework the `cfg` flags.